### ISSUE TITLE
[U.S.PPPM] notebook 살펴보기

### ISSUE THREAD

**Issue by heehehe** [heehehe/kaggle#9]
- [x] https://www.kaggle.com/code/jhoward/getting-started-with-nlp-for-absolute-beginners
- [x] https://www.kaggle.com/code/jhoward/iterate-like-a-grandmaster
- [ ] https://www.kaggle.com/code/venkatkumar001/nlp-starter-almost-all-basic-concept
- [ ] https://www.kaggle.com/code/alincijov/nlp-starter-continuous-bag-of-words-cbow
- [ ] https://www.kaggle.com/code/remekkinas/eda-and-feature-engineering
- [ ] https://www.kaggle.com/code/yasufuminakama/pppm-deberta-v3-large-baseline-w-w-b-train
https://www.kaggle.com/code/yasufuminakama/pppm-deberta-v3-large-baseline-inference
- [ ] https://www.kaggle.com/code/ksork6s4/uspppm-bert-for-patents-baseline-train
https://www.kaggle.com/code/ksork6s4/uspppm-bert-for-patents-baseline-inference

**Comment by heehehe** [comment]
## [jhoward/getting-started-with-nlp-for-absolute-beginners](https://www.kaggle.com/code/jhoward/getting-started-with-nlp-for-absolute-beginners)

<details><summary>kaggle tip</summary>

- kaggle에서 돌리고 있는건지 확인하는 방법
  ```python
  import os
  iskaggle = os.environ.get('KAGGLE_KERNEL_RUN_TYPE', '')
  ```
- kaggle api
  ```python
  from pathlib import Path
  
  creds = '{"username":"xxx","key":"xxx"}'
  cred_path = Path('~/.kaggle/kaggle.json').expanduser()
  if not cred_path.exists():
      cred_path.parent.mkdir(exist_ok=True)
      cred_path.write_text(creds)
      cred_path.chmod(0o600)

  path = Path('us-patent-phrase-to-phrase-matching')
  
  if not iskaggle and not path.exists():
      import zipfile, kaggle
      kaggle.api.competition_download_cli(str(path))
      zipfile.ZipFile(f'{path}.zip').extractall(path)
  
  if iskaggle:
    path = Path('../input/us-patent-phrase-to-phrase-matching')
    ! pip install -q datasets
  ```

</details>

- task 목적
  - patent phrase matching 문제를 classification 문제로 풀기
    - phrase 유사도 측정 : identical(1) / similar(0.5) / different(0)
- EDA (train.csv)
  - id - a unique identifier for a pair of phrases
  - anchor - 첫번째 phrase
  - target - 두번째 phrase
  - context - [CPC classification (version 2021.05)](https://en.wikipedia.org/wiki/Cooperative_Patent_Classification) (patent classification system으로, 유사한 주제들끼리 분류)
  - score - 유사도 (manual expert ratings에 의해 측정됨)  
  - 
      ```python
      df.describe(include='object')
      ```
      ![image](https://user-images.githubusercontent.com/41580746/171995175-6b07729e-9b63-43ad-8167-25743c5e11eb.png)
  - 
      ```python
      df['input'] = 'TEXT1: ' + df.context + '; TEXT2: ' + df.target + '; ANC1: ' + df.anchor
      ```
- tokenize
  - tokenization : text를 word 또는 token 단위로 쪼개기
    numericalization : 각 word 또는 token을 number로 변환
  - Transformers가 dataset 저장 위해 `Dataset` object 사용
      ```python
      from datasets import Dataset,DatasetDict
      ds = Dataset.from_pandas(df) 
      """
      Dataset({
          features: ['id', 'anchor', 'target', 'context', 'score', 'input'],
          num_rows: 36473
      })
      """
      ```
  - model 선정 : 대부분의 NLP 문제들이 starting point에서 사용하는 모델인 [DeBERTa: Decoding-enhanced BERT with Disentangled Attention](https://arxiv.org/abs/2006.03654) 이용
    (초기에는 small로 하고, 실제 실행에서는 large로 이용)
      ```python
      model_nm = 'microsoft/deberta-v3-small'
      from transformers import AutoModelForSequenceClassification,AutoTokenizer
      tokz = AutoTokenizer.from_pretrained(model_nm)
      
      ## 단어 시작은 "▁"로 되어 tokenize됨
      tokz.tokenize("G'day folks, I'm Jeremy from fast.ai!")  # ['▁G', "'", 'day', '▁folks', ',', '▁I', "'", 'm', '▁Jeremy', '▁from', '▁fast', '.', 'ai', '!']
      ```
      ```python
      def tok_func(x): return tokz(x["input"])
      
      tok_ds = ds.map(tok_func, batched=True)
      row = tok_ds[0]
      
      row['input']  # ('TEXT1: A47; TEXT2: abatement of pollution; ANC1: abatement',
      row['input_ids']  # [1, 54453, 435, 294, 336, ... ]
      
      tokz.vocab['▁of']  # '▁of' 에 대한 unique integer : 265
      
      tok_ds = tok_ds.rename_columns({'score':'labels'})
      ```
- test / validation sets
  - train 데이터를 75:25 비율의 train:validation set으로 쪼개기
      ```python
      dds = tok_ds.train_test_split(0.25, seed=42)
      """
      DatasetDict({
          train: Dataset({
              features: ['id', 'anchor', 'target', 'context', 'labels', 'input', 'input_ids', 'token_type_ids', 'attention_mask'],
              num_rows: 27354
          })
          test: Dataset({
              features: ['id', 'anchor', 'target', 'context', 'labels', 'input', 'input_ids', 'token_type_ids', 'attention_mask'],
              num_rows: 9119
          })
      })
      """
      ```
  - test.csv
      ```python
      eval_df = pd.read_csv(path/'test.csv')
      eval_df.describe()
      ```
      ![image](https://user-images.githubusercontent.com/41580746/172012503-48e82e18-323e-481a-89e2-adb69221fe16.png)
      ```python
      eval_df['input'] = 'TEXT1: ' + eval_df.context + '; TEXT2: ' + eval_df.target + '; ANC1: ' + eval_df.anchor
      eval_ds = Dataset.from_pandas(eval_df).map(tok_func, batched=True)
      ```
- metrics
  - 예측값과 실제값 similarity score 간의 Pearson correlation coefficient
      ```python
      def corr(x,y): return np.corrcoef(x,y)[0][1]
      
      def show_corr(df, a, b):
          x,y = df[a],df[b]
          plt.scatter(x,y, alpha=0.5, s=4)
          plt.title(f'{a} vs {b}; r: {corr(x, y):.2f}')
          
      def corr_d(eval_pred): return {'pearson': corr(*eval_pred)}
      ```
- training
  - model train
     ```python
     from transformers import TrainingArguments,Trainer
     
     bs = 128  # batch size
     epochs = 4
     lr = 8e-5  # learning rate
     
     args = TrainingArguments('outputs', 
                     learning_rate=lr, warmup_ratio=0.1, lr_scheduler_type='cosine', fp16=True,
                     evaluation_strategy="epoch", per_device_train_batch_size=bs, per_device_eval_batch_size=bs*2,
                     num_train_epochs=epochs, weight_decay=0.01, report_to='none'
                   )
     model = AutoModelForSequenceClassification.from_pretrained(model_nm, num_labels=1)
     trainer = Trainer(model, args, train_dataset=dds['train'], eval_dataset=dds['test'], 
                       tokenizer=tokz, compute_metrics=corr_d)
                       
     trainer.train()
     preds = trainer.predict(eval_ds).predictions.astype(float)
     preds = np.clip(preds, 0, 1)
     ```
  - submission 파일 생성
     ```python
     import datasets
     submission = datasets.Dataset.from_dict({
        'id': eval_ds['id'],
        'score': preds
     })
    
    submission.to_csv('submission.csv', index=False)
    ```

**Comment by heehehe** [comment]
## [jhoward/iterate-like-a-grandmaster](https://www.kaggle.com/code/jhoward/iterate-like-a-grandmaster)

- import & EDA
  - initialize
    ```python
    from fastai.imports import *   # np, pd, plt 등 basic imports 제공
    if iskaggle: path = Path('../input/us-patent-phrase-to-phrase-matching')
    ```
  - train / test
    ```python
    df = pd.read_csv(path/'train.csv')
    eval_df = pd.read_csv(path/'test.csv')
    ```
  - anchor (첫번째 phrase)
    ```python
    df.anchor.value_counts()
    """
    component composite coating              152
    sheet supply roller                      150
    source voltage                           140
    perfluoroalkyl group                     136
    el display                               135
                                            ... 
    plug nozzle                                2
    shannon                                    2
    dry coating composition1                   2
    peripheral nervous system stimulation      1
    conduct conducting material                1
    Name: anchor, Length: 733, dtype: int64
    """
    ```
  - target (두번째 phrase)
    ```python
    df.target.value_counts()
    """
    composition                    24
    data                           22
    metal                          22
    motor                          22
    assembly                       21
                                   ..
    switching switch over valve     1
    switching switch off valve      1
    switching over valve            1
    switching off valve             1
    wooden substrate                1
    Name: target, Length: 29340, dtype: int64
    """
    ```
  - context (CPC classification)
    ```python
    df.context.value_counts()
    """
    H01    2186
    H04    2177
    G01    1812
    A61    1477
    F16    1091
           ... 
    B03      47
    F17      33
    B31      24
    A62      23
    F26      18
    Name: context, Length: 106, dtype: int64
    """
    ```
  - context의 첫번째 분류에 대한 value_counts
    ```python
    df['section'] = df.context.str[0]
    df.section.value_counts()
    """
    B    8019
    H    6195
    G    6013
    C    5288
    A    4094
    F    4054
    E    1531
    D    1279
    Name: section, dtype: int64
    """
    ```
  - score (유사도 점수)
    ```python
    df.score.hist()
    ```
    ![image](https://user-images.githubusercontent.com/41580746/172014000-cfa6057e-d5db-4f92-8c9f-bd29b34d552f.png)
    ```python
    df[df.score==1]
    ```
    ![image](https://user-images.githubusercontent.com/41580746/172014037-26e298bb-9cc0-4ba0-9d93-b65f78a22dc2.png)
- initial setting
  - import
    ```python
    from torch.utils.data import DataLoader
    import warnings, transformers, logging, torch
    from transformers import TrainingArguments,Trainer
    from transformers import AutoModelForSequenceClassification,AutoTokenizer
    import datasets
    from datasets import load_dataset, Dataset, DatasetDict
    
    ## Huggingface가 warnings 과도하게 많이 출력해서 ignore 하도록
    warnings.simplefilter('ignore')
    logging.disable
    ```
  - tokenizer
    ```python
    ## 2분 이내로 돌릴 수 있으면서 reasonable한 정확도 얻을 수 있는 모델로 선정
    model_nm = 'microsoft/deberta-v3-small'
    tokz = AutoTokenizer.from_pretrained(model_nm)
    
    ## special separator 를 통해 context, anchor, target을 결합
    sep = tokz.sep_token   # '[SEP]'
    df['inputs'] = df.context + sep + df.anchor + sep + df.target
    
    ## dataframe을 huggingface dataset으로 변환
    ds = Dataset.from_pandas(df).rename_column('score', 'label')
    eval_ds = Dataset.from_pandas(eval_df)
    
    ## tokenize 위한 함수 (Dataset.map에서 사용하기 위함)
    def tok_func(x): return tokz(x['inputs'])
    tok_func(ds[0])
    """
    {'input_ids': [1, 336, 5753, 2, 47284, 2, 47284, 265, 6435, 2], 
     'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 
     'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}
    """
    ## input_ids에서 1: '[CLS]'(start of text), 2: '[SEP]'(separator token)
    tokz.all_special_tokens  # ['[CLS]', '[SEP]', '[UNK]', '[PAD]', '[MASK]']
    
    ## input tokenize + 불필요한 columns 제거
    inps = 'anchor','target','context'
    tok_ds = ds.map(tok_func, batched=True, remove_columns=inps+('inputs','id','section'))
    tok_ds[0]
    """
    {'label': 0.5,
     'input_ids': [1, 336, 5753, 2, 47284, 2, 47284, 265, 6435, 2],
     'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
     'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}
    """
    ```
- train / validation set
  - split data
    ```python
    anchors = df.anchor.unique()
    np.random.seed(42)
    np.random.shuffle(anchors)
    anchors[:5]   # array(['time digital signal', 'antiatherosclerotic', 'filled interior', 'dispersed powder', 'locking formation'], dtype=object)
    
    val_prop = 0.25
    val_sz = int(len(anchors) * val_prop)
    val_anchors = anchors[:val_sz]
    
    is_val = np.isin(df.anchor, val_anchors)
    idxs = np.arange(len(df))
    val_idxs = idxs[is_val]
    trn_idxs = idxs[~is_val]
    len(val_idxs), len(trn_idxs)   # (9116, 27357)
    
    dds = DatasetDict( {"train": tok_ds.select(trn_idxs), "test": tok_ds.select(val_idxs)} )
    
    df.iloc[trn_idxs].score.mean(), df.iloc[val_idxs].score.mean()  # 0.3623... , 0.3613....
    ```
- model initialization
  - metric & hyperparameter
    ```python
    def corr(eval_pred):
        return {'pearson': np.corrcoef(*eval_pred)[0][1]}
    
    ## hyperparameter 설정 (learning rate, batch size, weight decay, epochs)
    lr, bs, wd, epochs = 8e-5, 128, 0.01, 4
    ```
  - 모델 돌릴 함수 정의
    ```python
    ## arguments 설정
    args = TrainingArguments('outputs', 
        learning_rate=lr, 
        warmup_ratio=0.1, 
        lr_scheduler_type='cosine', 
        fp16=True, 
        evaluation_strategy="epoch", 
        per_device_train_batch_size=bs, 
        per_device_eval_batch_size=bs*2,
        num_train_epochs=epochs,
        weight_decay=wd,
        report_to='none')
    
    ## train
    model = AutoModelForSequenceClassification.from_pretrained(model_nm, num_labels=1)
    trainer = Trainer(model, args, 
        train_dataset=dds['train'], eval_dataset=dds['test'], 
        tokenizer=tokz, compute_metrics=corr)
    trainer.train()
    ```
    ![image](https://user-images.githubusercontent.com/41580746/172117544-6b744bb9-7a10-485e-81c6-7a1ba4715e6d.png)
- improving the model
  - 데이터 생성 및 train 함수 등 정의
    ```python
    def get_dds(df):
        ds = Dataset.from_pandas(df).rename_column('score', 'label')
        tok_ds = ds.map(tok_func, batched=True, remove_columns(inps+('inputs','id','section'))
        return DatasetDict( {"train": tok_ds.select(trn_idxs), "test": tok_ds.select(val_idxs)} )
        
    def get_model():
        return AutoModelForSequenceClassification.from_pretrained(model_nm, num_labels=1)
    
    def get_trainer(dds, model=None):
        if model is None: model = get_model()
        args = TrainingArguments('outputs', learning_rate=lr, warmup_ratio=0.1,
                        lr_scheduler_type='cosine', fp16=True, evaluation_strategy="epoch", 
                        per_device_train_batch_size=bs, per_device_eval_batch_size=bs*2,
                        num_train_epochs=epochs, weight_decay=wd, report_to='none')
        return Trainer(model, args, train_dataset=dds['train'], eval_dataset=dds['test'],
                        tokenizer=tokz, compute_metrics=corr)
    ```
  - special separator 변경 후 결과 확인 (약간 더 좋아져서 해당 방법 사용)
    ```python
    sep = ' [s] '
    df['inputs'] = df.context + sep + df.anchor + sep + df.target
    
    dds = get_dds(df)
    get_trainer(dds).train()
    ```
    ![image](https://user-images.githubusercontent.com/41580746/172119210-4b95a83c-52ce-4b77-bb36-7bcb70b46237.png)
  - lowercase로 변환 (separator 변환보다는 덜하지만, 초기보단 좋은 결과 나와서 사용)
    ```python
    df['inputs'] = df.inputs.str.lower()
    dds = get_dds(df)
    get_trainer(dds).train()
    ```
    ![image](https://user-images.githubusercontent.com/41580746/172119560-5663011b-0bfb-4386-9cd4-259c642bd8b7.png)
  - patent section에 대한 special token 생성
    ```python
    df['sectok'] = '[%s]' % df.section
    sectoks = list(df.sectok.unique())
    tokz.add_special_tokens( {'additional_special_tokens': sectoks} )
    
    df['inputs'] = df.sectok + sep + df.context + sep + df.anchor.str.lower() + sep + df.target
    dds = get_dds(df)
    
    ## 여러 token들을 추가했으므로 embedding matrix을 resize 
    model = get_model()
    model.resize_token_embeddings(len(tokz))
    
    trainer = get_trainer(dds, model=model)
    trainer.train()
    ```
    ![image](https://user-images.githubusercontent.com/41580746/172120119-9f26ec99-283e-41df-b2b8-421979117f42.png)
  - Cross Validation
    ```python
    from sklearn.model_selection import StratifiedGroupKFold

    n_folds = 4
    cv = StratifiedGroupKFold(n_splits=n_folds)
    
    df = df.sample(frac=1, random_state=42)
    scores = (df.score * 100).astype(int)
    folds = list(cv.split(idxs, scores, df.anchor))
    """
    [(array([    0,     1,     2, ..., 36469, 36471, 36472]),
      array([    8,    13,    14, ..., 36453, 36464, 36470])),
     (array([    0,     1,     5, ..., 36470, 36471, 36472]),
      array([    2,     3,     4, ..., 36459, 36461, 36462])),
     (array([    1,     2,     3, ..., 36467, 36470, 36472]),
      array([    0,     7,    11, ..., 36468, 36469, 36471])),
     (array([    0,     2,     3, ..., 36469, 36470, 36471]),
      array([    1,     5,     9, ..., 36465, 36467, 36472]))]
    """
    
    def get_fold(folds, fold_num):
        trn, val = folds[fold_num]
        return DatasetDict( {"train": tok_ds.select(trn), "test": tok_ds.select(val)} )
    
    dds = get_fold(folds, 0)
    """
    DatasetDict({
        train: Dataset({
            features: ['label', 'input_ids', 'token_type_ids', 'attention_mask'],
            num_rows: 27346
        })
        test: Dataset({
            features: ['label', 'input_ids', 'token_type_ids', 'attention_mask'],
            num_rows: 9127
        })
    })
    """
    
    metrics = [o['eval_pearson'] for o in trainer.state.log_history if 'eval_pearson' in o]
    metrics[-1]   # 0.8201...
    ```
